sim/src/image.rs

```diff
@@ -1337,6 +1356,55 @@ fn verify_trailer(flash: &SimMultiFlash, slot: &SlotInfo,
     !failed
 }
 
+/// Install a partition table.  This is a simplified partition table that
+/// we write at the beginning of flash so make it easier for external tools
+/// to analyze these images.
+fn install_ptable(flash: &mut SimMultiFlash, areadesc: &AreaDesc) {
+    let ids: HashSet<u8> = areadesc.iter_areas().map(|area| area.device_id).collect();
+    for &id in &ids {
+        // If there are any partitions in this device that start at 0, and
+        // aren't marked as the BootLoader partition, avoid adding the
+        // partition table.  This makes it harder to view the image, but
+        // avoids messing up images already written.
+        if areadesc.iter_areas().any(|area| {
+            area.device_id == id &&
+                area.off == 0 &&
+                area.flash_id != FlashId::BootLoader
+        }) {
+            let special: Vec<FlashId> = areadesc.iter_areas()
+                .filter(|area| area.device_id == id && area.off == 0)
+                .map(|area| area.flash_id)
+                .collect();
```

Comment:
Wouldn't it be possible to avoid iterating twice, by just collecting the `flash_ids` and checking if there is any?

Reply:
I could be, but I'm not sure which would be better. I probably should pub the `let special ...` in an if statement based on whether info logging is even enabled, since right now we compute the whole thing, and then immediately discard it in the normal case of there not being a message printed.

The first `any` iteration will stop as soon as it finds something matching the criteria. Even in the case of printing, it is probably faster to just do that instead of allocation a Vec to hold the result.

I'll push a version with the conditional.